server/src/main/java/org/eclipse/lsp/cobol/core/engine/CobolLanguageEngine.java

```diff
@@ -157,7 +157,42 @@ CobolParser getCobolParser(CommonTokenStream tokens) {
 
   @NonNull
   private Function<SyntaxError, SyntaxError> convertError(@NonNull Map<Token, Locality> mapping) {
-    return err -> err.toBuilder().locality(mapping.get(err.getOffendedToken())).build();
+    return err ->
+        err.toBuilder()
+            .locality(
+                mapping.getOrDefault(
+                    err.getOffendedToken(), getNearestLocality(err.getOffendedToken(), mapping)))
+            .suggestion(convertErrorMessage(err.getSuggestion()))
+            .build();
+  }
+
+  private Locality getNearestLocality(Token referenceToken, @NonNull Map<Token, Locality> mapping) {
+    if (Objects.isNull(referenceToken)) return null;
+    Optional<Locality> nearestToken =
+        mapping.entrySet().stream()
+            .filter(
+                entry ->
+                    StringUtils.normalizeSpace(entry.getKey().getText())
+                            .contains(StringUtils.normalizeSpace(referenceToken.getText()))
+                        && entry.getKey().getCharPositionInLine()
+                            == referenceToken.getCharPositionInLine()
+                        && entry.getKey().getLine() == referenceToken.getLine()
+                        && entry.getKey().getStartIndex() <= referenceToken.getStartIndex()
+                        && entry.getKey().getStopIndex() >= referenceToken.getStopIndex())
+            .map(Map.Entry::getValue)
+            .findFirst();
+
+    return nearestToken.orElse(null); // null is safe here.
+  }
+
+  private String convertErrorMessage(String suggestion) {
```

Comment:
I would rather catch the exceptions on the service side

Reply:
Make sense. Done.